### PR TITLE
[SPARK-35809][PYTHON] Add `index_col` argument for ps.sql

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes adding an argument `index_col` for `ps.sql` function, to preserve the index when users want.

NOTE that the `reset_index()` have to be performed before using `ps.sql` with `index_col`.

```python
>>> psdf
   A  B
a  1  4
b  2  5
c  3  6
>>> psdf_reset_index = psdf.reset_index()
>>> ps.sql("SELECT * from {psdf_reset_index} WHERE A > 1", index_col="index")
       A  B
index
b      2  5
c      3  6
```

Otherwise, the index is always lost.

```python
>>> ps.sql("SELECT * from {psdf} WHERE A > 1")
   A  B
0  2  5
1  3  6
```

### Why are the changes needed?

Index is one of the key object for the existing pandas users, so we should provide the way to keep the index after computing the `ps.sql`.

### Does this PR introduce _any_ user-facing change?

Yes, the new argument is added.

### How was this patch tested?

Add a unit test and manually check the build pass.

### CODE REVIEW

File: python/pyspark/pandas/sql_processor.py [hunk 1]
```diff
@@ -65,6 +66,9 @@ def sql(
     ----------
     query : str
         the SQL query
+    index_col: str or list of str, optional
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/sql_processor.py:69]: is `list of str` the way in NumPy documentation style?

Comment (itholic) [reply python/pyspark/pandas/sql_processor.py:69]: Yeah I think so. It's documented in style guide for numpydoc.

Comment (itholic) [reply python/pyspark/pandas/sql_processor.py]: Oh, but the space before `:` is missing. I just added the space.

File: python/pyspark/pandas/sql_processor.py [hunk 2]
```diff
@@ -65,6 +66,9 @@ def sql(
     ----------
     query : str
         the SQL query
+    index_col : str or list of str, optional
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/sql_processor.py:69]: I mean I don't know the standard way of describing the list of strings (`list of str`). Are you able to find any example like this in pandas or NumPy documentation?

File: python/pyspark/pandas/sql_processor.py [hunk 3]
```diff
@@ -252,17 +265,21 @@ def execute(self) -> DataFrame:
...
         finally:
             for v in self._temp_views:
                 self._session.catalog.dropTempView(v)
-        return DataFrame(sdf)
 
-    def _convert(self, key: str) -> Any:
+        if index_col is None:
+            return DataFrame(sdf)
+        else:
+            return DataFrame(sdf).set_index(index_col)
```
Comment (HyukjinKwon) [inline python/pyspark/pandas/sql_processor.py:280]: can we update internal instead? see read_spark_io as an example

File: python/pyspark/pandas/sql_processor.py [hunk 4]
```diff
@@ -252,17 +265,21 @@ def execute(self) -> DataFrame:
         res = ""
         try:
             for (pre, inner, _, _) in blocks:
-                var_next = "" if inner is None else self._convert(inner)
+                var_next = "" if inner is None else self._convert(inner, index_col)
```
Comment (ueshin) [inline python/pyspark/pandas/sql_processor.py:268]: I don't think we should handle inputs.
We could have a few DataFrames for the sql command, and those can have different levels of Index.

Comment (itholic) [reply python/pyspark/pandas/sql_processor.py]: Oh... yeah I misunderstood this meaning before.

Just addressed it, thanks!

**Comment by ueshin** [comment]
@itholic Could you also update the description?

**Review by HyukjinKwon — APPROVED** [review]

**Comment by itholic** [comment]
> @itholic Could you also update the description?

Just updated. Thanks

**Comment by itholic** [comment]
Let me fix the doctest

**Comment by HyukjinKwon** [comment]
Merged to master and branch-3.2.

**Pull request closed without merging**